In [1]:
# Supress Warnings

import warnings
warnings.filterwarnings('ignore')

# Import the numpy and pandas packages

import numpy as np
import pandas as pd

## Loading and Inspection of Datasets

We first check the encodings of the companies, rounds2 csv datasets. 
We then import the data using appropriate encoding and store them in two Dataframes called `companies` and `rounds2`, respectively.

The encoding of each of the csv dataset can be checked by executing the code (now commented out) in the cell below this one. 
The encoding types were observed to be as follows:

1. companies.txt : Encoding = 'Windows-1254'
2. rounds.csv : Encoding = 'Windows-1254'

**However**, using encoding as '__Windows-1254__' (as detected using chardet.detect) does **not** load the datasets properly.
On further observation, it was found that the datasets have multiple encodings in them and using encoding "__ISO-8859-1__" reads the entire dataset successfully. 

The below code has been commented as eventually we are not using the encoding obtained by these steps.

In [2]:
# I. Check encoding
# Import the chardet package to check file encoding
#import chardet

# 1. 'companies.txt'
#companies_enc = chardet.detect(open("companies.txt",'rb').read())
#print("\nEncoding details of 'companies.txt' file :\n",companies_enc)

# 2. 'rounds2.csv'
#rounds2_enc = chardet.detect(open("rounds2.csv",'rb').read())
#print("\nEncoding details of 'rounds2.csv' file :\n",rounds2_enc)

In [3]:
# We now import the two datasets into the dataframes companies and rounds2, respectively. 

# On inspection of the file, it was found that the delimiter is a tab, ie. '\t' character.
companies = pd.read_csv("companies.txt" , sep="\t",encoding = "ISO-8859-1")

# On inspection of the file, it was found that the delimiter is a semi-colon, ie. ';' character.
rounds2 = pd.read_csv("rounds2.csv" , sep=",",encoding = "ISO-8859-1")


In [4]:
# There were still some encoding errors observed, such as those in the row identified by index 33892.
#This was due to the fact that multiple encoding types were used in the dataset.
print("\nFor 'companies' Dataframe:-")
print("\nBefore handling encoding issues:\n")
display(companies.iloc[33892])

# Thus we change the encoding of 'company_permalink' column from 'ISO-8859-1' to 'ascii' to address the encoding issue.
companies['permalink'] = companies['permalink'].str.encode('ISO-8859-1').str.decode('ascii', 'ignore')

# Check for the same index 33892 to ensure that the encoding issues have been taken care of
print("\nAfter handling encoding issues:\n")
print("\nThe encoding issues related to the column 'name' can be ignored as we will not be using this column for analysis.")
display(companies.iloc[33892])

# Check first few rows of the cleaned dataframe
print("\nInitial 5 rows of cleaned dataframe:\n")
companies.head()


For 'companies' Dataframe:-

Before handling encoding issues:



permalink                   /Organization/Magnet-Tech-Ç£Çÿ³Ç§‘Æš€
name                                      Magnet Tech 磁石科技
homepage_url                                    http://www.buga.cn
category_list    Communications Hardware|Families|Hardware + So...
status                                                      closed
country_code                                                   NaN
state_code                                                     NaN
region                                                         NaN
city                                                           NaN
founded_at                                              02-05-2014
Name: 33892, dtype: object


After handling encoding issues:


The encoding issues related to the column 'name' can be ignored as we will not be using this column for analysis.


permalink                               /Organization/Magnet-Tech-
name                                      Magnet Tech 磁石科技
homepage_url                                    http://www.buga.cn
category_list    Communications Hardware|Families|Hardware + So...
status                                                      closed
country_code                                                   NaN
state_code                                                     NaN
region                                                         NaN
city                                                           NaN
founded_at                                              02-05-2014
Name: 33892, dtype: object


Initial 5 rows of cleaned dataframe:



permalink                    name  \
0                  /Organization/-Fame                   #fame   
1               /Organization/-Qounter                :Qounter   
2  /Organization/-The-One-Of-Them-Inc-  (THE) ONE of THEM,Inc.   
3                /Organization/0-6-Com                 0-6.com   
4       /Organization/004-Technologies        004 Technologies   

                        homepage_url  \
0                 http://livfame.com   
1             http://www.qounter.com   
2                http://oneofthem.jp   
3                 http://www.0-6.com   
4  http://004gmbh.de/en/004-interact   

                                       category_list     status country_code  \
0                                              Media  operating          IND   
1  Application Platforms|Real Time|Social Network...  operating          USA   
2                                  Apps|Games|Mobile  operating          NaN   
3                                        Curated Web  operating          CHN   
4                                           Software  operating          USA   

  state_code                 region           city  founded_at  
0         16                 Mumbai         Mumbai         NaN  
1         DE             DE - Other  Delaware City  04-09-2014  
2        NaN                    NaN            NaN         NaN  
3         22                Beijing        Beijing  01-01-2007  
4         IL  Springfield, Illinois      Champaign  01-01-2010

In [5]:
# There are still some encoding errors as seen in the row identified by index 58473

print("\nFor 'rounds2' Dataframe:-")
print("\nBefore handling encoding issues:\n")
display(rounds2.loc[58473])

# Changing the encoding of 'company_permalink' column from 'ISO-8859-1' to 'ascii' to address encoding issues
rounds2['company_permalink'] = rounds2['company_permalink'].str.encode('ISO-8859-1').str.decode('ascii', 'ignore')

print("\nAfter handling encoding issues:\n")
display(rounds2.loc[58473])

# Check first few rows of the cleaned dataframe
print("\nInitial 5 rows of cleaned dataframe:\n")
rounds2.head()


For 'rounds2' Dataframe:-

Before handling encoding issues:



company_permalink                   /ORGANIZATION/MAGNET-TECH-Ç£ÇŸ³Ç§‘ÆŠ€
funding_round_permalink    /funding-round/8fc91fbb32bc95e97f151dd0cb4166bf
funding_round_type                                                    seed
funding_round_code                                                     NaN
funded_at                                                       16-08-2014
raised_amount_usd                                              1.62558e+06
Name: 58473, dtype: object


After handling encoding issues:



company_permalink                               /ORGANIZATION/MAGNET-TECH-
funding_round_permalink    /funding-round/8fc91fbb32bc95e97f151dd0cb4166bf
funding_round_type                                                    seed
funding_round_code                                                     NaN
funded_at                                                       16-08-2014
raised_amount_usd                                              1.62558e+06
Name: 58473, dtype: object


Initial 5 rows of cleaned dataframe:



company_permalink  \
0                  /organization/-fame   
1               /ORGANIZATION/-QOUNTER   
2               /organization/-qounter   
3  /ORGANIZATION/-THE-ONE-OF-THEM-INC-   
4                /organization/0-6-com   

                           funding_round_permalink funding_round_type  \
0  /funding-round/9a01d05418af9f794eebff7ace91f638            venture   
1  /funding-round/22dacff496eb7acb2b901dec1dfe5633            venture   
2  /funding-round/b44fbb94153f6cdef13083530bb48030               seed   
3  /funding-round/650b8f704416801069bb178a1418776b            venture   
4  /funding-round/5727accaeaa57461bd22a9bdd945382d            venture   

  funding_round_code   funded_at  raised_amount_usd  
0                  B  05-01-2015         10000000.0  
1                  A  14-10-2014                NaN  
2                NaN  01-03-2014           700000.0  
3                  B  30-01-2014          3406878.0  
4                  A  19-03-2008          2000000.0

-  ### We now Inspect both the dataframe for their columns, shapes, variable types etc. 

In [6]:
# For 'companies' dataframe

# Print the number of Rows and Columns in 'companies' dataframe
print("\nThe 'companies' dataframe has {} Rows and {} Columns.\n".format(*companies.shape))

# To understand the various attributes and their type in the 'companies' dataframe.
print("\nAttributes of 'companies' dataframe:\n")
companies.info() 

# Display the first few rows of the 'companies' dataframe to get a feel of the data
print("\nInitial 5 rows of 'companies' dataframe:\n")
display(companies.head())

# Column-wise null percentages to get a feel of missing data
print("\nColumn-wise null percentages to get a feel of missing data : \n{}" \
      .format(round(100*(companies.isnull().sum()/len(companies.index)), 2)))


The 'companies' dataframe has 66368 Rows and 10 Columns.


Attributes of 'companies' dataframe:

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66368 entries, 0 to 66367
Data columns (total 10 columns):
permalink        66368 non-null object
name             66367 non-null object
homepage_url     61310 non-null object
category_list    63220 non-null object
status           66368 non-null object
country_code     59410 non-null object
state_code       57821 non-null object
region           58338 non-null object
city             58340 non-null object
founded_at       51147 non-null object
dtypes: object(10)
memory usage: 5.1+ MB

Initial 5 rows of 'companies' dataframe:



permalink                    name  \
0                  /Organization/-Fame                   #fame   
1               /Organization/-Qounter                :Qounter   
2  /Organization/-The-One-Of-Them-Inc-  (THE) ONE of THEM,Inc.   
3                /Organization/0-6-Com                 0-6.com   
4       /Organization/004-Technologies        004 Technologies   

                        homepage_url  \
0                 http://livfame.com   
1             http://www.qounter.com   
2                http://oneofthem.jp   
3                 http://www.0-6.com   
4  http://004gmbh.de/en/004-interact   

                                       category_list     status country_code  \
0                                              Media  operating          IND   
1  Application Platforms|Real Time|Social Network...  operating          USA   
2                                  Apps|Games|Mobile  operating          NaN   
3                                        Curated Web  operating          CHN   
4                                           Software  operating          USA   

  state_code                 region           city  founded_at  
0         16                 Mumbai         Mumbai         NaN  
1         DE             DE - Other  Delaware City  04-09-2014  
2        NaN                    NaN            NaN         NaN  
3         22                Beijing        Beijing  01-01-2007  
4         IL  Springfield, Illinois      Champaign  01-01-2010


Column-wise null percentages to get a feel of missing data : 
permalink         0.00
name              0.00
homepage_url      7.62
category_list     4.74
status            0.00
country_code     10.48
state_code       12.88
region           12.10
city             12.10
founded_at       22.93
dtype: float64


In [7]:
# For 'rounds2' dataframe

# Print the number of Rows and Columns in 'rounds2' dataframe
print("\nThe 'rounds2' dataframe has {} Rows and {} Columns.\n".format(*rounds2.shape))

# To understand the various attributes and their type in the 'rounds2' dataframe.
print("\nAttributes of 'rounds2' dataframe:\n")
rounds2.info() 

# Display the first few rows of the 'rounds2' dataframe to get a feel of the data
print("\nInitial 5 rows of 'rounds2' dataframe:\n")
display(rounds2.head())

# Column-wise null percentages to get a feel of missing data
print("\nColumn-wise null percentages to get a feel of missing data : \n{}" \
      .format(round(100*(rounds2.isnull().sum()/len(rounds2.index)), 2)))


The 'rounds2' dataframe has 114949 Rows and 6 Columns.


Attributes of 'rounds2' dataframe:

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 114949 entries, 0 to 114948
Data columns (total 6 columns):
company_permalink          114949 non-null object
funding_round_permalink    114949 non-null object
funding_round_type         114949 non-null object
funding_round_code         31140 non-null object
funded_at                  114949 non-null object
raised_amount_usd          94959 non-null float64
dtypes: float64(1), object(5)
memory usage: 5.3+ MB

Initial 5 rows of 'rounds2' dataframe:



company_permalink  \
0                  /organization/-fame   
1               /ORGANIZATION/-QOUNTER   
2               /organization/-qounter   
3  /ORGANIZATION/-THE-ONE-OF-THEM-INC-   
4                /organization/0-6-com   

                           funding_round_permalink funding_round_type  \
0  /funding-round/9a01d05418af9f794eebff7ace91f638            venture   
1  /funding-round/22dacff496eb7acb2b901dec1dfe5633            venture   
2  /funding-round/b44fbb94153f6cdef13083530bb48030               seed   
3  /funding-round/650b8f704416801069bb178a1418776b            venture   
4  /funding-round/5727accaeaa57461bd22a9bdd945382d            venture   

  funding_round_code   funded_at  raised_amount_usd  
0                  B  05-01-2015         10000000.0  
1                  A  14-10-2014                NaN  
2                NaN  01-03-2014           700000.0  
3                  B  30-01-2014          3406878.0  
4                  A  19-03-2008          2000000.0


Column-wise null percentages to get a feel of missing data : 
company_permalink           0.00
funding_round_permalink     0.00
funding_round_type          0.00
funding_round_code         72.91
funded_at                   0.00
raised_amount_usd          17.39
dtype: float64


## Data Cleaning

In [8]:
# We now do some Standardization and cleaning of the data.

# For 'companies' dataframe:-

# Since the data in 'category_list' is title-cased, convert to a standard format to lower case
# This would help us in merging data with 'mapping' dataframe later
companies['category_list'] = companies['category_list'].str.lower()

# Since the 'permalink' column has data in upper and lower case, convert to a standard format to lower case
# This would help us in merging data with 'rounds2' dataframe later
companies['permalink'] = companies['permalink'].str.lower()


# For 'rounds2' dataframe:-

# Drop 'funding_round_code' column as it is insignificant to further analysis.
rounds2.drop(columns='funding_round_code', axis=1, inplace=True)

# Since the 'permalink' column has data in upper and lower case, convert to a standard format to lower case.
# This would help us in merging data with 'companies' dataframe later.
rounds2['company_permalink'] = rounds2['company_permalink'].str.lower()


- ### Table 1.1 Results

Q. How many unique companies are present in rounds2?

In [9]:
#We check the length of the pandas series containing all the unique values from the company_permalink column of rounds2 dataframe.
#We use the unique() function for this.
len(rounds2.company_permalink.unique())

66368

Therefore, we see that there are **66,368 unique** companies in **"rounds2"** Data-Set. 

Q. How many unique companies are present in the companies file?

In [10]:
#We check the length of the pandas series containing all the unique values from the permalink column of companies dataframe.
#We use the unique() function for this.
len(companies.permalink.unique())

66368

Therefore, we see that there are **66,368 unique** companies in the **"companies"** Data-Set.

Q. In the companies data frame, which column can be used as the unique key for each company? Write the name of the column.

In [11]:
companies.head()

permalink                    name  \
0                  /organization/-fame                   #fame   
1               /organization/-qounter                :Qounter   
2  /organization/-the-one-of-them-inc-  (THE) ONE of THEM,Inc.   
3                /organization/0-6-com                 0-6.com   
4       /organization/004-technologies        004 Technologies   

                        homepage_url  \
0                 http://livfame.com   
1             http://www.qounter.com   
2                http://oneofthem.jp   
3                 http://www.0-6.com   
4  http://004gmbh.de/en/004-interact   

                                       category_list     status country_code  \
0                                              media  operating          IND   
1  application platforms|real time|social network...  operating          USA   
2                                  apps|games|mobile  operating          NaN   
3                                        curated web  operating          CHN   
4                                           software  operating          USA   

  state_code                 region           city  founded_at  
0         16                 Mumbai         Mumbai         NaN  
1         DE             DE - Other  Delaware City  04-09-2014  
2        NaN                    NaN            NaN         NaN  
3         22                Beijing        Beijing  01-01-2007  
4         IL  Springfield, Illinois      Champaign  01-01-2010

We can see that, the column named **"permalink"** column can be used as a unique key for each company. Answer: **Permalink**

Q. Are there any companies in the rounds2 file which are not  present in companies ? Answer Y/N.

In [12]:
#We check the length of the pandas series got after taking the NEGATION of all the values of 
#company_permalink column (rounds2 dataframe) present in the permalink column(companies dataframe).
#We use the unique() function for this.
len(rounds2[~rounds2['company_permalink'].isin(companies['permalink'])])

0

Therefore, we can see that there are **NO** companies in the rounds2 dataset that are not there in the companies dataset.
Answer: **N**

In [13]:
#We now merge the companies and rounds2 dataframes and create a merged dataframe called master_frame.
#We use left join to do this.
master_frame = rounds2.merge(companies,how='left',left_on=['company_permalink'],right_on=['permalink'])
master_frame.head()

company_permalink  \
0                  /organization/-fame   
1               /organization/-qounter   
2               /organization/-qounter   
3  /organization/-the-one-of-them-inc-   
4                /organization/0-6-com   

                           funding_round_permalink funding_round_type  \
0  /funding-round/9a01d05418af9f794eebff7ace91f638            venture   
1  /funding-round/22dacff496eb7acb2b901dec1dfe5633            venture   
2  /funding-round/b44fbb94153f6cdef13083530bb48030               seed   
3  /funding-round/650b8f704416801069bb178a1418776b            venture   
4  /funding-round/5727accaeaa57461bd22a9bdd945382d            venture   

    funded_at  raised_amount_usd                            permalink  \
0  05-01-2015         10000000.0                  /organization/-fame   
1  14-10-2014                NaN               /organization/-qounter   
2  01-03-2014           700000.0               /organization/-qounter   
3  30-01-2014          3406878.0  /organization/-the-one-of-them-inc-   
4  19-03-2008          2000000.0                /organization/0-6-com   

                     name            homepage_url  \
0                   #fame      http://livfame.com   
1                :Qounter  http://www.qounter.com   
2                :Qounter  http://www.qounter.com   
3  (THE) ONE of THEM,Inc.     http://oneofthem.jp   
4                 0-6.com      http://www.0-6.com   

                                       category_list     status country_code  \
0                                              media  operating          IND   
1  application platforms|real time|social network...  operating          USA   
2  application platforms|real time|social network...  operating          USA   
3                                  apps|games|mobile  operating          NaN   
4                                        curated web  operating          CHN   

  state_code      region           city  founded_at  
0         16      Mumbai         Mumbai         NaN  
1         DE  DE - Other  Delaware City  04-09-2014  
2         DE  DE - Other  Delaware City  04-09-2014  
3        NaN         NaN            NaN         NaN  
4         22     Beijing        Beijing  01-01-2007

Q.How many observations are present in master_frame?

In [14]:
len(master_frame)

114949

Therefore, we see **master_frame** dataframe has **114949** entries. Answer: **114949**

In [15]:
# Now, We clean the Master Frame.

# Dropping the redundant column 'permalink' as it has the same data as the 'company_permalink' column
master_frame.drop('permalink',axis=1,inplace=True)

# Remove rows which have 'raised_amount_usd' = 'NaN' as these are forming a large percentage of missing data.
# Further, these rows are insignificant to our analysis as we are interested in those companies which have received funding.
# Also, it is better to remove these rows in the initial part itself so that there is less overhead in subsequent steps.
master_frame = master_frame[~master_frame.raised_amount_usd.isna()]

# Convert the unit of the 'raised_amount_usd' columns from `$` to `million $`.
# 1 million = 1000000. 
# Divide by 1000000 to convert unit of 'raised_amount_usd' column to million $ and assign back to same columns.
master_frame['raised_amount_usd'] = master_frame['raised_amount_usd']/1000000

# Remove rows which have 'country_code' = 'NaN' as these are forming a large percentage of missing data.
# A 'NaN' value in 'country_code' indicates that either we do not know the country of investment or
# the country_code field was not captured for that investment.
# In either case, these rows become insignificant as the investment country is a vital information for our analysis.
master_frame = master_frame[~master_frame.country_code.isna()]


## Checkpoint 2: Funding Type Analysis

In [16]:
#We now group the masterframe by the funding type and check the aggregate sum,count and mean of each funding type. 
master_frame.groupby('funding_round_type')['raised_amount_usd'].agg(['sum','count', 'mean'])

sum  count        mean
funding_round_type                                     
angel                    4287.814720   4427    0.968560
convertible_note         1766.149564   1326    1.331938
debt_financing         112278.774046   6533   17.186403
equity_crowdfunding       580.436907   1136    0.510948
grant                    8922.267135   1979    4.508473
non_equity_assistance      29.701089     63    0.471446
post_ipo_debt           25587.220256    151  169.451790
post_ipo_equity         39677.295350    601   66.018794
private_equity         135090.064238   1835   73.618564
product_crowdfunding      494.574424    332    1.489682
secondary_market         2364.288903     29   81.527204
seed                    15884.502810  21233    0.748104
undisclosed             21430.658506   1352   15.851079
venture                564620.108462  48111   11.735780

**TABLE 2.1 Results**

Average funding amount of venture type : *__11.735780 million USD__*

Average funding amount of angel type : __0.968560 million USD__

Average funding amount of seed type : __0.748104 million USD__

Average funding amount of private equity type : __73.618564 million USD__

Considering that Spark Funds wants to invest between 5 to 15 million USD per investment round,
which investment type is the most suitable for it? 

Answer : __'venture'__

## Checkpoint 3: Country Analysis

In [17]:
#We now create a 'top9' dataframe, based on the total investment amount each country has received for the 'venture' type funding

# master_frame['funding_round_type'] == 'venture' --> select only those rows for which 'funding_round_type' is 'venture'
# groupby('country_code')['raised_amount_usd'].sum() : groups the data by country_code and calculates the sum() based on group
# sort_values(ascending=False) : sorts the resultant in descending order
# reset_index().head(9) : resets index and select only top 9 Rows
top9 = pd.DataFrame(master_frame[master_frame['funding_round_type'] == 'venture'].\
            groupby('country_code')['raised_amount_usd'].sum().\
            sort_values(ascending=False)).reset_index().head(9)

In [18]:
# Display the 'top9' dataframe
top9

country_code  raised_amount_usd
0          USA      422510.842796
1          CHN       39835.418773
2          GBR       20245.627416
3          IND       14391.858718
4          CAN        9583.332317
5          FRA        7259.536732
6          ISR        6907.514579
7          DEU        6346.959822
8          JPN        3363.676611

Referring to the 'Countries_where_English_is_an_official_language.pdf' file provided and by manual inspection, we can conclude the below

**TABLE 3.1 Results**

#### Analysing the Top 3 English-Speaking Countries

1.Top English-speaking country : __United States of America (USA)__

2.Second English-speaking country : __United Kingdom (GBR)__

3.Third English-speaking country  : __India (IND)__

Note: We skip China (which was at 2nd position) since it is **NOT** an English Speaking Country.

## Checkpoint 4: Sector Analysis 1

In [19]:
#We first convert the datatype of every element of the category_list column to "string" datatype
master_frame['category_list'] = master_frame['category_list'].astype(str)

#We now Extract the primary sector of each category list from the category_list column of the master dataframe
#and store it in a new column called, "primary_sector"
#We use the apply function to apply spliting of each value, by the pipe character, of the category_list series.
#and take only the first substring.

master_frame['primary_sector'] = master_frame['category_list'].apply(lambda x: x.split('|')[0])

#We check the master_frame to see wheather primary_sector column has been succesfully created or not.
master_frame.head()

company_permalink  \
0                  /organization/-fame   
2               /organization/-qounter   
4                /organization/0-6-com   
6     /organization/01games-technology   
7  /organization/0ndine-biomedical-inc   

                           funding_round_permalink funding_round_type  \
0  /funding-round/9a01d05418af9f794eebff7ace91f638            venture   
2  /funding-round/b44fbb94153f6cdef13083530bb48030               seed   
4  /funding-round/5727accaeaa57461bd22a9bdd945382d            venture   
6  /funding-round/7d53696f2b4f607a2f2a8cbb83d01839        undisclosed   
7  /funding-round/2b9d3ac293d5cdccbecff5c8cb0f327d               seed   

    funded_at  raised_amount_usd                    name  \
0  05-01-2015           10.00000                   #fame   
2  01-03-2014            0.70000                :Qounter   
4  19-03-2008            2.00000                 0-6.com   
6  01-07-2014            0.04125      01Games Technology   
7  11-09-2009            0.04336  Ondine Biomedical Inc.   

             homepage_url                                      category_list  \
0      http://livfame.com                                              media   
2  http://www.qounter.com  application platforms|real time|social network...   
4      http://www.0-6.com                                        curated web   
6  http://www.01games.hk/                                              games   
7    http://ondinebio.com                                      biotechnology   

      status country_code state_code      region           city  founded_at  \
0  operating          IND         16      Mumbai         Mumbai         NaN   
2  operating          USA         DE  DE - Other  Delaware City  04-09-2014   
4  operating          CHN         22     Beijing        Beijing  01-01-2007   
6  operating          HKG        NaN   Hong Kong      Hong Kong         NaN   
7  operating          CAN         BC   Vancouver      Vancouver  01-01-1997   

          primary_sector  
0                  media  
2  application platforms  
4            curated web  
6                  games  
7          biotechnology

**Loading the mapping.csv to mapping dataframe**

In [20]:
# We now Import the 'mapping.csv' dataset into 'mapping' dataframe.
# On inspection of the file, it was found that the delimiter is a comma, ie. ',' character.
mapping = pd.read_csv("mapping.csv" , sep=",",encoding = "ascii")

**We now Inspect the mapping dataframe.**

In [21]:
# Print the number of Rows and Columns in 'mapping' dataframe
print("\nThe 'mapping' dataframe has {} Rows and {} Columns.\n".format(*mapping.shape))

# To understand the various attributes and their type in the 'mapping' dataframe.
print("\nAttributes of 'mapping' dataframe:\n")
mapping.info() 

# Display the first few rows of the 'mapping' dataframe to get a feel of the data
print("\nInitial 5 rows of 'mapping' dataframe:\n")
display(mapping.head())

# Column-wise null percentages to get a feel of missing data
print("\nColumn-wise null percentages to get a feel of missing data : \n{}" \
      .format(round(100*(mapping.isnull().sum()/len(mapping.index)), 2)))


The 'mapping' dataframe has 688 Rows and 10 Columns.


Attributes of 'mapping' dataframe:

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 688 entries, 0 to 687
Data columns (total 10 columns):
category_list                              687 non-null object
Automotive & Sports                        688 non-null int64
Blanks                                     688 non-null int64
Cleantech / Semiconductors                 688 non-null int64
Entertainment                              688 non-null int64
Health                                     688 non-null int64
Manufacturing                              688 non-null int64
News, Search and Messaging                 688 non-null int64
Others                                     688 non-null int64
Social, Finance, Analytics, Advertising    688 non-null int64
dtypes: int64(9), object(1)
memory usage: 53.8+ KB

Initial 5 rows of 'mapping' dataframe:



category_list  Automotive & Sports  Blanks  Cleantech / Semiconductors  \
0            NaN                    0       1                           0   
1             3D                    0       0                           0   
2    3D Printing                    0       0                           0   
3  3D Technology                    0       0                           0   
4     Accounting                    0       0                           0   

   Entertainment  Health  Manufacturing  News, Search and Messaging  Others  \
0              0       0              0                           0       0   
1              0       0              1                           0       0   
2              0       0              1                           0       0   
3              0       0              1                           0       0   
4              0       0              0                           0       0   

   Social, Finance, Analytics, Advertising  
0                                        0  
1                                        0  
2                                        0  
3                                        0  
4                                        1


Column-wise null percentages to get a feel of missing data : 
category_list                              0.15
Automotive & Sports                        0.00
Blanks                                     0.00
Cleantech / Semiconductors                 0.00
Entertainment                              0.00
Health                                     0.00
Manufacturing                              0.00
News, Search and Messaging                 0.00
Others                                     0.00
Social, Finance, Analytics, Advertising    0.00
dtype: float64


**Data Cleaning**

In [22]:
#We first drop the 'Blanks' column entirely, since 'Blanks' (entries in the master_frame where there is NO primary sector) 
#is NOT to be considered as one of the Eight Main Sectors for our Analysis.
mapping.drop('Blanks', axis=1, inplace=True)

#We check wheather 'Blanks' column has been successfully dropped or not.
mapping.head()

category_list  Automotive & Sports  Cleantech / Semiconductors  \
0            NaN                    0                           0   
1             3D                    0                           0   
2    3D Printing                    0                           0   
3  3D Technology                    0                           0   
4     Accounting                    0                           0   

   Entertainment  Health  Manufacturing  News, Search and Messaging  Others  \
0              0       0              0                           0       0   
1              0       0              1                           0       0   
2              0       0              1                           0       0   
3              0       0              1                           0       0   
4              0       0              0                           0       0   

   Social, Finance, Analytics, Advertising  
0                                        0  
1                                        0  
2                                        0  
3                                        0  
4                                        1

In [23]:
#We now use the melt() function of pandas to “Unpivot” the frame, where our Id variable would be the category_list column.
#and all other column headers (ie. the Eight Main sectors) would be our variables.
#This would generate for us a dataframe where each value of the category list would have a corresponding value of 0 or 1
#for each column header(Main Sectors), in row-wise format.
#var_name='main_sector' renames the column name of 'variables' to 'main_sector'

melted_frame = mapping.melt(id_vars=['category_list'], var_name='main_sector')

#We check the melted frame
melted_frame.head()

category_list          main_sector  value
0            NaN  Automotive & Sports      0
1             3D  Automotive & Sports      0
2    3D Printing  Automotive & Sports      0
3  3D Technology  Automotive & Sports      0
4     Accounting  Automotive & Sports      0

In [24]:
#We now remove all rows where value is 0. 
#Since we need the mapping, retaining only the values of 1 would leave us with a frame where each category would be successfully
#mapped with its corresponding "main_sector"
melted_frame = melted_frame[melted_frame.value==1]

#We again check the melted frame
melted_frame.head()

category_list          main_sector  value
8   Adventure Travel  Automotive & Sports      1
14         Aerospace  Automotive & Sports      1
45              Auto  Automotive & Sports      1
46   Automated Kiosk  Automotive & Sports      1
47        Automotive  Automotive & Sports      1

**On further inspecting the 'category_list' column of the melted dataframe we find that "Na" has been replaced with 0**
It appears that the substring **'Na'** has been replaced by string **'0'** (zero) due to some unknown reason. ( So, **'0notechnology'** should have been **'nanotechnology'**, **'0tural language processing'** should have been **'natural language processing'** and so on.
This only occurs in the mapping dataframe, while the master_frame dataframe does not show signs of this issue.

In [25]:
display(mapping.loc[(243,275,471,534),['category_list']])

category_list
243       Enterprise 2.0
275   Fi0ncial Exchanges
471        0notechnology
534  Predictive A0lytics

**So, Now we rectify this data issue in the melted dataframe.**

In [26]:
# We first convert the datatype of all values of the category_list column to string datatype and also to lowercase. 
melted_frame['category_list'] = melted_frame['category_list'].astype(str).str.lower()

#Since, directly replacing All "0" with "na" in the category list column would lead us to modify or replace "0"
#in strings Where there was genuinely 0. (which shouldnt be replaced). for example, "Enterprise 2.0"

#To take care of this, we create a new column called 'primary_sector' which would contain all the values of the category_list, 
#BUT, here ALL "0" would be replaced with "na"
melted_frame['primary_sector'] = melted_frame['category_list'].apply(lambda x: x.replace("0","na"))

#So, Now we have 2 columns, one called "category_list", Where NO "0" has been replaced with "na"
#and, another called "primary_sector", Where ALL "0" has been replaced with "na"

#Now, we simply slice off the both these columns, along with their corresponding "main_sector"
#and concatenate them, one on top of the other.
c1=melted_frame[['category_list','main_sector']].rename({"category_list":"primary_sector"}, axis='columns')
c2=melted_frame[['primary_sector','main_sector']]

#After dropping the duplicates, we are left with a mapping of primary sector to the "main_sector",
mapping = pd.concat([c1, c2], ignore_index=True).drop_duplicates()

mapping.head()

primary_sector          main_sector
0  adventure travel  Automotive & Sports
1         aerospace  Automotive & Sports
2              auto  Automotive & Sports
3   automated kiosk  Automotive & Sports
4        automotive  Automotive & Sports

This mapping would contain category_list values with genuine "0" as well as with "0" replaced by "na"
for example, it would have **"enterprise 2.0" as well as "enterprise 2.na"** and **"predictive a0lytics" as well as "predictive analytics"** mapped to the same main_sector.
As a result, we have a mapping which could be used by the primary_category of
the master_frame to fetch its main_sector.
**Note:** We use this method because **otherwise** we would have to **manually** check all the entries where "0" occurs, verify if it is a genuine 0 or has been mistakenly replaced (in the dataset) instead of "na", and then devise a **hard-coded** method which would replace the 0 with na selectively and therby omit modifying entry with genuine 0.
Since this would be **implausible** with datasets having entries (containing 0) to the tune of several thousands, we have used the **concatenation method** described in the previous cell. 

In [27]:
#We now merge the master_frame with the mapping_frame.
# Code to merge the 'master_frame' dataframe using key as 'primary_sector' column with 
#                   'mapping' dataframe using key as 'primary_sector' column.
# An 'inner' join is done so as to choose only those records from 'master_frame' whose 'primary_sector' 
# is present in the 'mapping' dataframe's 'primary_sector'
master_frame = master_frame.merge(mapping,how='inner',on='primary_sector')


## Checkpoint 4 Results

In [28]:
# Check first few rows of the merged dataframe for inclusion of 'primary_sector' and 'main_sector' columns
master_frame.head()

company_permalink  \
0           /organization/-fame   
1           /organization/90min   
2           /organization/90min   
3           /organization/90min   
4  /organization/a-dance-for-me   

                           funding_round_permalink   funding_round_type  \
0  /funding-round/9a01d05418af9f794eebff7ace91f638              venture   
1  /funding-round/21a2cbf6f2fb2a1c2a61e04bf930dfe6              venture   
2  /funding-round/bd626ed022f5c66574b1afe234f3c90d              venture   
3  /funding-round/fd4b15e8c97ee2ffc0acccdbe1a98810              venture   
4  /funding-round/9ab9dbd17bf010c79d8415b2c22be6fa  equity_crowdfunding   

    funded_at  raised_amount_usd            name                 homepage_url  \
0  05-01-2015              10.00           #fame           http://livfame.com   
1  06-10-2015              15.00           90min         http://www.90min.com   
2  07-05-2013               5.80           90min         http://www.90min.com   
3  26-03-2014              18.00           90min         http://www.90min.com   
4  26-03-2014               1.09  A Dance for Me  http://www.adanceforme.com/   

                         category_list     status country_code state_code  \
0                                media  operating          IND         16   
1  media|news|publishing|soccer|sports  operating          GBR         H9   
2  media|news|publishing|soccer|sports  operating          GBR         H9   
3  media|news|publishing|soccer|sports  operating          GBR         H9   
4       media|news|photo sharing|video  operating          USA         MT   

     region      city  founded_at primary_sector    main_sector  
0    Mumbai    Mumbai         NaN          media  Entertainment  
1    London    London  01-01-2011          media  Entertainment  
2    London    London  01-01-2011          media  Entertainment  
3    London    London  01-01-2011          media  Entertainment  
4  Missoula  Missoula  31-07-2011          media  Entertainment

## Checkpoint 5: Sector Analysis 2

In [29]:
#We create a function, that would return a sub-dataframe containing 
#observations of funding type 'venture' falling within the 5-15 million USD range,
#for any given input of countrycode.
def get_subframe(cntry_code):
    return master_frame[(master_frame.country_code==cntry_code) 
                        & (master_frame.funding_round_type=="venture") 
                        & (master_frame.raised_amount_usd>=5) & (master_frame.raised_amount_usd<=15) ]


#We create the 3 dataframes D1,D2,D3 (using the above function) for the top 3 english countries found in our earlier analysis. 
D1 = get_subframe("USA")
D2 = get_subframe("GBR")
D3 = get_subframe("IND")

print("\nFirst 5 rows of 'D1' dataframe:\n")
display(D1.head())

print("\nFirst 5 rows of 'D2' dataframe:\n")
display(D2.head())

print("\nFirst 5 rows of 'D3' dataframe:\n")
display(D3.head())


First 5 rows of 'D1' dataframe:



company_permalink  \
7   /organization/all-def-digital   
31       /organization/chefs-feed   
61   /organization/huffingtonpost   
62   /organization/huffingtonpost   
85        /organization/matchmine   

                            funding_round_permalink funding_round_type  \
7   /funding-round/452a2342fe720285c3b92e9bd927d9ba            venture   
31  /funding-round/adca195749ae9ace84684723fbe75e5b            venture   
61  /funding-round/7f05940c4d2dfecb8e50a0e5720e5065            venture   
62  /funding-round/9241ae16e08df17ebdc064e49e23035a            venture   
85  /funding-round/41ac526630da57ad6eb9d02431b17657            venture   

     funded_at  raised_amount_usd                 name  \
7   06-08-2014                5.0      All Def Digital   
31  26-02-2015                5.0            ChefsFeed   
61  01-08-2006                5.0  The Huffington Post   
62  01-09-2007                5.0  The Huffington Post   
85  01-09-2007               10.0            MatchMine   

                     homepage_url                           category_list  \
7        http://alldefdigital.com                                   media   
31       http://www.chefsfeed.com     media|mobile|restaurants|technology   
61  http://www.huffingtonpost.com                   media|news|publishing   
62  http://www.huffingtonpost.com                   media|news|publishing   
85           http://matchmine.com  media|news|reviews and recommendations   

       status country_code state_code         region           city  \
7   operating          USA         CA    Los Angeles    Los Angeles   
31  operating          USA         CA    SF Bay Area  San Francisco   
61   acquired          USA         NY  New York City       New York   
62   acquired          USA         NY  New York City       New York   
85     closed          USA         MA         Boston        Needham   

    founded_at primary_sector    main_sector  
7          NaN          media  Entertainment  
31  01-01-2012          media  Entertainment  
61  09-05-2005          media  Entertainment  
62  09-05-2005          media  Entertainment  
85  01-01-2007          media  Entertainment


First 5 rows of 'D2' dataframe:



company_permalink  \
1                          /organization/90min   
2                          /organization/90min   
225  /organization/common-interest-communities   
257                /organization/geospock-ltd-   
258                /organization/geospock-ltd-   

                             funding_round_permalink funding_round_type  \
1    /funding-round/21a2cbf6f2fb2a1c2a61e04bf930dfe6            venture   
2    /funding-round/bd626ed022f5c66574b1afe234f3c90d            venture   
225  /funding-round/8195587cbd5e51af7514ee92ef4ba6ba            venture   
257  /funding-round/cf3fe3b7c86186b9f478d0ea37613f7a            venture   
258  /funding-round/e5e4ef4ebae63fc36ef0cd57dd20ff1c            venture   

      funded_at  raised_amount_usd                         name  \
1    06-10-2015              15.00                        90min   
2    07-05-2013               5.80                        90min   
225  09-07-2014              10.00  Common Interest Communities   
257  01-10-2014               5.46                GeoSpock Ltd.   
258  05-10-2015               5.40                GeoSpock Ltd.   

                              homepage_url  \
1                     http://www.90min.com   
2                     http://www.90min.com   
225  http://commoninterestcommunities.com/   
257                http://www.geospock.com   
258                http://www.geospock.com   

                                        category_list     status country_code  \
1                 media|news|publishing|soccer|sports  operating          GBR   
2                 media|news|publishing|soccer|sports  operating          GBR   
225  application platforms|internet|software|startups  operating          GBR   
257         application platforms|databases|real time  operating          GBR   
258         application platforms|databases|real time  operating          GBR   

    state_code  region       city  founded_at         primary_sector  \
1           H9  London     London  01-01-2011                  media   
2           H9  London     London  01-01-2011                  media   
225         H9  London     London         NaN  application platforms   
257         C3  London  Cambridge  01-01-2013  application platforms   
258         C3  London  Cambridge  01-01-2013  application platforms   

                    main_sector  
1                 Entertainment  
2                 Entertainment  
225  News, Search and Messaging  
257  News, Search and Messaging  
258  News, Search and Messaging


First 5 rows of 'D3' dataframe:



company_permalink  \
0               /organization/-fame   
550           /organization/babajob   
593  /organization/bharat-matrimony   
640     /organization/bluestone-com   
642     /organization/bluestone-com   

                             funding_round_permalink funding_round_type  \
0    /funding-round/9a01d05418af9f794eebff7ace91f638            venture   
550  /funding-round/b72eaac5ea12ac0f50573ac3d6d46b8d            venture   
593  /funding-round/e37673bc7b0f1dfd3782f8f7abdb9ec8            venture   
640  /funding-round/452a7fc1f34df2d3dcda4e28234bc671            venture   
642  /funding-round/f5b252d6442ce231bb01586ca1821f63            venture   

      funded_at  raised_amount_usd              name  \
0    05-01-2015              10.00             #fame   
550  28-04-2015              10.00           Babajob   
593  05-02-2008              11.75  Bharat Matrimony   
640  24-01-2012               5.00     Bluestone.com   
642  18-03-2014              10.00     Bluestone.com   

                       homepage_url  \
0                http://livfame.com   
550          http://www.babajob.com   
593  http://www.bharatmatrimony.com   
640            http://bluestone.com   
642            http://bluestone.com   

                                         category_list     status  \
0                                                media  operating   
550  curated web|information technology|services|st...  operating   
593                           curated web|match-making  operating   
640                                        curated web  operating   
642                                        curated web  operating   

    country_code state_code     region       city  founded_at primary_sector  \
0            IND         16     Mumbai     Mumbai         NaN          media   
550          IND         19  Bangalore  Bangalore  27-08-2007    curated web   
593          IND         25    Chennai    Chennai  12-03-1969    curated web   
640          IND         19  Bangalore  Bangalore  01-01-2011    curated web   
642          IND         19  Bangalore  Bangalore  01-01-2011    curated web   

                    main_sector  
0                 Entertainment  
550  News, Search and Messaging  
593  News, Search and Messaging  
640  News, Search and Messaging  
642  News, Search and Messaging

## **Checkpoint 5 Results for Table 5.1 (Answers to Questions 1 to 10)** 

We first find all the answers (1 to 10) for the D1 dataframe, followed by D2 and then D3 dataframe.

In [30]:
#For D1:-
print("Total Amount of Investments: {} Million USD\n".format(D1.raised_amount_usd.sum()))

print("Total Number of Investments: {}\n".format(len(D1.index)))

print("Note:- 'Sum' column in the frame is in million USD.")
display(D1.groupby("main_sector")['raised_amount_usd'].agg(['count','sum']).sort_values('count',ascending=False))

print("Thus, we can conculde. (For D1)")
print("Based on count of investments.\n")

print("Top Sector is 'Others' with total count of 2950.")
print("Second Best Sector is 'Social, Finance, Analytics, Advertising' with total count of 2714.")
print("Third Best Sector is 'Cleantech / Semiconductors' with total count of 2350.")


print("\nCompany with Highest Investment in TopSector (ie. 'Others'):")
display(D1[D1.main_sector == 'Others'][['company_permalink', 'name', 'raised_amount_usd']].\
        groupby(['company_permalink','name'])['raised_amount_usd'].sum().\
        sort_values(ascending=False).to_frame().head(1))

print("\nCompany with Highest Investment in SecondBest Sector (ie. 'Social, Finance, Analytics, Advertising'):")

display(D1[D1.main_sector == 'Social, Finance, Analytics, Advertising'][['company_permalink', 'name', 'raised_amount_usd']].\
        groupby(['company_permalink','name'])['raised_amount_usd'].sum().\
        sort_values(ascending=False).to_frame().head(1))


Total Amount of Investments: 107757.097294 Million USD

Total Number of Investments: 12063

Note:- 'Sum' column in the frame is in million USD.


count           sum
main_sector                                                 
Others                                    2950  26321.007002
Social, Finance, Analytics, Advertising   2714  23807.376964
Cleantech / Semiconductors                2350  21633.430822
News, Search and Messaging                1583  13971.567428
Health                                     909   8211.859357
Manufacturing                              799   7258.553378
Entertainment                              591   5099.197982
Automotive & Sports                        167   1454.104361

Thus, we can conculde. (For D1)
Based on count of investments.

Top Sector is 'Others' with total count of 2950.
Second Best Sector is 'Social, Finance, Analytics, Advertising' with total count of 2714.
Third Best Sector is 'Cleantech / Semiconductors' with total count of 2350.

Company with Highest Investment in TopSector (ie. 'Others'):


raised_amount_usd
company_permalink         name                          
/organization/virtustream Virtustream               64.3


Company with Highest Investment in SecondBest Sector (ie. 'Social, Finance, Analytics, Advertising'):


raised_amount_usd
company_permalink         name                                              
/organization/shotspotter SST Inc. (Formerly ShotSpotter)          67.933006

In [31]:
#For D2:-
print("Total Amount of Investments: {} Million USD\n".format(D2.raised_amount_usd.sum()))

print("Total Number of Investments: {}\n".format(len(D2.index)))

print("Note:- 'Sum' column in the frame is in million USD.")

display(D2.groupby("main_sector")['raised_amount_usd'].agg(['count','sum']).sort_values('count',ascending=False))

print("Thus, we can conculde. (For D2)")
print("Based on count of investments.\n")

print("Top Sector is 'Others' with total count of 147.")
print("Second Best Sector is 'Social, Finance, Analytics, Advertising' with total count of 133.")
print("Third Best Sector is 'Cleantech / Semiconductors' with total count of 130.")

print("\nCompany with Highest Investment in TopSector (ie. 'Others'):")
display(D2[D2.main_sector == 'Others'][['company_permalink', 'name', 'raised_amount_usd']].\
        groupby(['company_permalink','name'])['raised_amount_usd'].sum().\
        sort_values(ascending=False).to_frame().head(1))

print("\nCompany with Highest Investment in SecondBest Sector (ie. 'Social, Finance, Analytics, Advertising'):")

display(D2[D2.main_sector == 'Social, Finance, Analytics, Advertising'][['company_permalink', 'name', 'raised_amount_usd']].\
        groupby(['company_permalink','name'])['raised_amount_usd'].sum().\
        sort_values(ascending=False).to_frame().head(1))


Total Amount of Investments: 5379.078691000001 Million USD

Total Number of Investments: 621

Note:- 'Sum' column in the frame is in million USD.


count          sum
main_sector                                                
Others                                     147  1283.624289
Social, Finance, Analytics, Advertising    133  1089.404014
Cleantech / Semiconductors                 130  1163.990056
News, Search and Messaging                  73   615.746235
Entertainment                               56   482.784687
Manufacturing                               42   361.940335
Health                                      24   214.537510
Automotive & Sports                         16   167.051565

Thus, we can conculde. (For D2)
Based on count of investments.

Top Sector is 'Others' with total count of 147.
Second Best Sector is 'Social, Finance, Analytics, Advertising' with total count of 133.
Third Best Sector is 'Cleantech / Semiconductors' with total count of 130.

Company with Highest Investment in TopSector (ie. 'Others'):


raised_amount_usd
company_permalink            name                             
/organization/electric-cloud Electric Cloud               37.0


Company with Highest Investment in SecondBest Sector (ie. 'Social, Finance, Analytics, Advertising'):


raised_amount_usd
company_permalink                   name                                    
/organization/celltick-technologies Celltick Technologies               37.5

In [32]:
#For D3:-
print("Total Amount of Investments: {} Million USD\n".format(D3.raised_amount_usd.sum()))

print("Total Number of Investments: {}\n".format(len(D3.index)))

print("Note:- 'Sum' column in frame is in million USD.")
display(D3.groupby("main_sector")['raised_amount_usd'].agg(['count','sum']).sort_values('count',ascending=False))

print("Thus, we can conculde. (For D3)")
print("Based on count of investments.\n")

print("Top Sector is 'Others' with total count of 110.")
print("Second Best Sector is 'Social, Finance, Analytics, Advertising' with total count of 60.")
print("Third Best Sector is 'News, Search and Messaging' with total count of 52.")


print("\nCompany with Highest Investment in TopSector (ie. 'Others'):")
display(D3[D3.main_sector == 'Others'][['company_permalink', 'name', 'raised_amount_usd']].\
        groupby(['company_permalink','name'])['raised_amount_usd'].sum().\
        sort_values(ascending=False).to_frame().head(1))

print("\nCompany with Highest Investment in SecondBest Sector (ie. 'Social, Finance, Analytics, Advertising'):")

display(D3[D3.main_sector == 'Social, Finance, Analytics, Advertising'][['company_permalink', 'name', 'raised_amount_usd']].\
        groupby(['company_permalink','name'])['raised_amount_usd'].sum().\
        sort_values(ascending=False).to_frame().head(1))


Total Amount of Investments: 2949.5436019999997 Million USD

Total Number of Investments: 328

Note:- 'Sum' column in frame is in million USD.


count          sum
main_sector                                                
Others                                     110  1013.409507
Social, Finance, Analytics, Advertising     60   550.549550
News, Search and Messaging                  52   433.834545
Entertainment                               33   280.830000
Manufacturing                               21   200.900000
Cleantech / Semiconductors                  20   165.380000
Health                                      19   167.740000
Automotive & Sports                         13   136.900000

Thus, we can conculde. (For D3)
Based on count of investments.

Top Sector is 'Others' with total count of 110.
Second Best Sector is 'Social, Finance, Analytics, Advertising' with total count of 60.
Third Best Sector is 'News, Search and Messaging' with total count of 52.

Company with Highest Investment in TopSector (ie. 'Others'):


raised_amount_usd
company_permalink          name                           
/organization/firstcry-com FirstCry.com               39.0


Company with Highest Investment in SecondBest Sector (ie. 'Social, Finance, Analytics, Advertising'):


raised_amount_usd
company_permalink             name                              
/organization/manthan-systems Manthan Systems               50.7

#### Exporting the final master_frame to CSV for  Inspections in Tableau

In [33]:
master_frame.to_csv("tableaudata_clean_masterfrmae.csv")

**THE END**